File: SpeedTest/implementations/javascript/mergesort.js
#!/usr/bin/env node

function mergeSort(arr) {
    if (arr.length <= 1) {
        return arr;
    }

    const mid = Math.floor(arr.length / 2);
    const left = mergeSort(arr.slice(0, mid));
    const right = mergeSort(arr.slice(mid));

    return merge(left, right);
}

function merge(left, right) {
    const result = [];
    let i = 0;
    let j = 0;

    while (i < left.length && j < right.length) {
        if (left[i] <= right[j]) {
            result.push(left[i]);
            i++;
        } else {
            result.push(right[j]);
            j++;
        }
    }

    return result.concat(left.slice(i)).concat(right.slice(j));
}

function generateRandomArray(size) {
    return Array.from({ length: size }, () => Math.floor(Math.random() * 1000) + 1);
}

function main() {
    // Run 100 times with different random arrays
    for (let i = 0; i < 100; i++) {
        const arr = generateRandomArray(1000);
        mergeSort(arr);
    }
}

main();

File: SpeedTest/implementations/python/mergesort.py
#!/usr/bin/env python3
import random
import time

def merge_sort(arr):
    if len(arr) <= 1:
        return arr

    mid = len(arr) // 2
    left = merge_sort(arr[:mid])
    right = merge_sort(arr[mid:])

    return merge(left, right)

def merge(left, right):
    result = []
    i = j = 0

    while i < len(left) and j < len(right):
        if left[i] <= right[j]:
            result.append(left[i])
            i += 1
        else:
            result.append(right[j])
            j += 1

    result.extend(left[i:])
    result.extend(right[j:])
    return result

def generate_random_array(size):
    return [random.randint(1, 1000) for _ in range(size)]

def main():
    # Run 100 times with different random arrays
    for _ in range(100):
        arr = generate_random_array(1000)
        merge_sort(arr)

if __name__ == "__main__":
    main()
